File: GAIL_simulator_py/src/domain/machine/entities/slot_machine.py
# src/domain/machine/entities/slot_machine.py
import logging
from typing import Dict, List, Any, Tuple, Optional

from .reel import Reel
from ..services.win_evaluation import WinEvaluator


class SlotMachine:
    """
    Represents a slot machine with its configuration, reels, and win evaluation logic.
    Core entity in the machine domain.
    """
    def __init__(self, machine_id: str, config: Dict[str, Any], rng_strategy=None):
        """
        Initialize the slot machine.
        
        Args:
            machine_id: Unique identifier for this machine
            config: Machine configuration dictionary
            rng_strategy: Random number generator strategy (optional)
        """
        self.id = machine_id
        self.logger = logging.getLogger(f"domain.machine.{machine_id}")
        self.logger.info(f"Initializing slot machine: {machine_id}")
        
        # Save original configuration
        self.config = config
        
        # Initialize with RNG strategy
        self.rng = rng_strategy
        
        # Extract basic parameters
        self.normal_symbols = config.get("symbols", {}).get("normal", list(range(11)))
        self.wild_symbols = config.get("symbols", {}).get("wild", [101])
        self.scatter_symbol = config.get("symbols", {}).get("scatter", 20)
        self.free_spins_count = config.get("free_spins", 10)
        self.free_spins_multiplier = config.get("free_spins_multiplier", 1)
        self.window_size = config.get("window_size", 3)
        
        # Load components
        self._load_reels(config.get("reels", {}))
        self._load_paylines(config.get("paylines", []))
        self._load_pay_table(config.get("pay_table", []))
        self._load_bet_table(config.get("bet_table", []))

        # Initialize Win Evaluator as instance
        self._evaluator = WinEvaluator(self)
        
        self.logger.info(f"Slot machine {machine_id} initialized successfully")
        
    def _load_reels(self, reels_config: Dict[str, Any]):
        """
        Load reel configurations.
        
        Args:
            reels_config: Dictionary of reel configurations
        """
        
        self.reels = {}
        
        # Process each reel set (normal mode, free spins mode, etc.)
        for reel_set_name, reel_set_data in reels_config.items():
            if not isinstance(reel_set_data, dict):
                self.logger.warning(f"Invalid reel set format {reel_set_name}: expected dict")
                continue
                
            # Skip comment entries
            if reel_set_name.startswith('_'):
                continue
                
            self.reels[reel_set_name] = {}
            
            # Process each reel in this set
            for reel_name, symbols in reel_set_data.items():
                if not isinstance(symbols, list):
                    self.logger.warning(f"Invalid reel format {reel_set_name}.{reel_name}: expected list")
                    continue
                
                self.reels[reel_set_name][reel_name] = Reel(symbols, f"{reel_set_name}_{reel_name}")
                self.logger.debug(f"Loaded reel {reel_set_name}.{reel_name} with {len(symbols)} symbols")
        
        # Verify at least one valid reel set
        if not self.reels or 'normal' not in self.reels:
            self.logger.warning("No valid reels found or missing 'normal' reel set")
            # Create minimal default reel set
            self.reels['normal'] = {
                f'reel{i}': Reel([0, 1, 2, 3, 4, 5], f'default_normal_reel{i}') 
                for i in range(1, 6) 
            }
            
    def _load_paylines(self, paylines_config: List[Dict[str, Any]]):
        """
        Load payline configurations.
        
        Args:
            paylines_config: List of payline configurations
        """
        self.paylines = []
        
        for i, entry in enumerate(paylines_config):
            if not isinstance(entry, dict) or 'indices' not in entry:
                self.logger.warning(f"Invalid payline entry at index {i}: missing 'indices'")
                continue
                
            indices = entry.get('indices', [])
            if not isinstance(indices, list) or len(indices) < 3:
                self.logger.warning(f"Invalid payline indices at index {i}: {indices}")
                continue
                
            self.paylines.append(indices)
        
        if not self.paylines:
            self.logger.warning("No valid paylines found, using defaults")
            # Create default paylines
            self.paylines = [
                [0, 1, 2, 3, 4],       # Top row
                [5, 6, 7, 8, 9],       # Middle row
                [10, 11, 12, 13, 14],  # Bottom row
            ]
            
    def _load_pay_table(self, pay_table_config: List[Dict[str, Any]]):
        """
        Load pay table configuration.
        
        Args:
            pay_table_config: List of pay table entries
        """
        self.pay_table = {}
        
        for entry in pay_table_config:
            if not isinstance(entry, dict) or 'symbol' not in entry or 'payouts' not in entry:
                self.logger.warning(f"Invalid pay table entry: {entry}")
                continue
                
            symbol = entry.get('symbol')
            payouts = entry.get('payouts', [])
            
            if not isinstance(payouts, list) or len(payouts) < 3:
                self.logger.warning(f"Invalid payouts for symbol {symbol}: {payouts}")
                continue
                
            self.pay_table[symbol] = payouts
            
        if not self.pay_table:
            self.logger.warning("No valid pay table entries, using defaults")
            # Create default pay table
            self.pay_table = {
                "0": [5, 20, 100],
                "1": [5, 20, 100],
                "2": [5, 20, 100],
                "3": [5, 20, 100],
                "4": [5, 20, 100],
                "5": [5, 20, 100],
                str(self.scatter_symbol): [5, 20, 100]
            }
            
    def _load_bet_table(self, bet_table_config: List[Dict[str, Any]]):
        """
        Load bet table configuration.
        
        Args:
            bet_table_config: List of bet table entries
        """
        self.bet_table = {}
        
        for entry in bet_table_config:
            if not isinstance(entry, dict) or 'currency' not in entry or 'bet_options' not in entry:
                self.logger.warning(f"Invalid bet table entry: {entry}")
                continue
                
            currency = entry.get('currency')
            bet_options = entry.get('bet_options', [])
            
            if not isinstance(bet_options, list) or not bet_options:
                self.logger.warning(f"Invalid bet options for currency {currency}: {bet_options}")
                continue
                
            self.bet_table[currency] = sorted(set(bet_options))  # Ensure options are unique and sorted
            
    def set_rng(self, rng_strategy):
        """
        Set or update the RNG strategy.
        
        Args:
            rng_strategy: RNG strategy instance
        """
        self.rng = rng_strategy
        self.logger.debug(f"Updated RNG strategy: {type(rng_strategy).__name__}")
        
    def spin(self, in_free: bool = False, num_free_left: int = 0) -> Tuple[List[int], bool, int]:
        """
        Execute a spin and return the result.
        
        Args:
            in_free: Whether in free spins mode
            num_free_left: Number of free spins remaining
            
        Returns:
            Tuple of (result_grid, trigger_free, num_free_left)
            - result_grid: Flattened array of symbols (row-major order)
            - trigger_free: Whether free spins were triggered
            - num_free_left: Number of free spins remaining
        """
        if self.rng is None:
            self.logger.error("No RNG strategy set, cannot spin")
            raise ValueError("No RNG strategy set for slot machine")
            
        # Determine which reel set to use
        reel_set_name = "bonus" if in_free else "normal"
        
        if reel_set_name not in self.reels:
            self.logger.warning(f"Reel set '{reel_set_name}' not found, using 'normal'")
            reel_set_name = "normal"
            
        current_reels = self.reels[reel_set_name]
        reel_names = sorted(current_reels.keys())
        num_reels = len(reel_names)
        
        # Initialize result grid
        result = [0] * (num_reels * self.window_size)  # 3 rows x num_reels
        
        # Spin each reel and get visible symbols
        for i, name in enumerate(reel_names):
            reel = current_reels[name]
            pos = self.rng.get_random_int(0, len(reel) - 1)
            symbols = reel.get_symbols_at_position(pos, self.window_size)
            
            # Store symbols in flattened grid
            for row in range(self.window_size):
                result[row * num_reels + i] = symbols[row]
        
        # Check for free spins trigger
        if not in_free:
            # Count columns with scatter symbol
            scatter_cols = sum(
                any(result[row * num_reels + col] == self.scatter_symbol for row in range(self.window_size))
                for col in range(num_reels)
            )
            trigger_free = scatter_cols >= 3
            num_free_left = self.free_spins_count if trigger_free else 0
        else:
            # In free spins mode, decrement counter
            num_free_left = max(0, num_free_left - 1)
            trigger_free = num_free_left > 0
            
        self.logger.debug(
            f"Spin result: {result}, trigger_free={trigger_free}, num_free_left={num_free_left}"
        )
            
        return result, trigger_free, num_free_left
        
    @property
    def evaluator(self):
        return self._evaluator

    def evaluate_win(self, grid: List[int], bet: float, in_free: bool = False, active_lines: int = None) -> Dict[str, Any]:
        return self._evaluator.evaluate_wins(grid, bet, in_free, active_lines)
    
        
    def reset_state(self):
        """
        重置机器状态以用于新会话。
        主要是重新初始化RNG以确保不同的随机序列。
        """
        # 重置RNG
        if hasattr(self, 'rng') and self.rng:
            if hasattr(self.rng, 'initialize'):
                self.rng.initialize()
            elif hasattr(self.rng, 'seed'):
                import random
                self.rng.seed(random.randint(0, 1000000))
        
        self.logger.debug(f"Machine {self.id} state reset")
        
        
    def get_info(self) -> Dict[str, Any]:
        """
        Get information about this machine.
        
        Returns:
            Dictionary with machine information
        """
        return {
            'id': self.id,
            'reel_sets': list(self.reels.keys()),
            'reel_lengths': {
                set_name: {reel_name: len(reel) for reel_name, reel in reel_set.items()}
                for set_name, reel_set in self.reels.items()
            },
            'num_paylines': len(self.paylines),
            'wild_symbols': self.wild_symbols,
            'scatter_symbol': self.scatter_symbol,
            'free_spins_award': self.free_spins_count,
            'free_spins_multiplier': self.free_spins_multiplier,
            'available_currencies': list(self.bet_table.keys())
        }

File: GAIL_simulator_cpp/src/machines/slot_machine.cpp
// src/machines/slot_machine.cpp
#include "slot_machine.h"
#include "../utils/logger.h"
#include <algorithm>

namespace SlotSimulator {

const BetOptions SlotMachine::empty_bet_options_;

SlotMachine::SlotMachine(const MachineConfig& config) 
    : config_(config), rng_(std::random_device{}()) {
    
    // 创建PayTable
    pay_table_ = std::make_unique<PayTable>(config_.pay_table, config_.paylines);
    
    // 创建ReelSet
    for (const auto& [reel_set_name, reels] : config_.reels) {
        reel_sets_[reel_set_name] = std::make_unique<ReelSet>(reels, config_.window_size);
    }
    
    LOG_INFO("SlotMachine " + config_.machine_id + " initialized with " + 
             std::to_string(reel_sets_.size()) + " reel sets", "SlotMachine");
}

SpinResult SlotMachine::Spin(float bet_amount, bool in_free_spins, int free_spins_remaining) {
    SpinResult result;
    result.bet_amount = bet_amount;
    result.in_free_spins = in_free_spins;
    result.timestamp = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::system_clock::now().time_since_epoch()).count() / 1000.0;
    
    // 选择轮盘集合
    std::string reel_set_name = in_free_spins ? "bonus" : "normal";
    if (reel_sets_.find(reel_set_name) == reel_sets_.end()) {
        reel_set_name = "normal";  // 回退到normal
    }
    
    // 生成旋转结果
    result.grid = GenerateSpinGrid(reel_set_name);
    
    // 计算中奖金额
    result.win_amount = CalculateWinAmount(result.grid, bet_amount, config_.active_lines);
    result.profit = result.win_amount - bet_amount;
    
    // 检查免费旋转触发
    if (!in_free_spins) {
        result.trigger_free_spins = CheckFreeSpinsTrigger(result.grid);
        result.free_spins_remaining = result.trigger_free_spins ? config_.free_spins_count : 0;
    } else {
        result.trigger_free_spins = false;
        result.free_spins_remaining = std::max(0, free_spins_remaining - 1);
        // 免费旋转倍数
        result.win_amount *= config_.free_spins_multiplier;
        result.profit = result.win_amount - bet_amount;
    }
    
    return result;
} = in_free_spins ? "bonus" : "normal";
    if (reel_sets_.find(reel_set_name) == reel_sets_.end()) {
        reel_set_name = "normal";  // 回退到normal
    }
    
    // 生成旋转结果
    result.grid = GenerateSpinGrid(reel_set_name);
    
    // 计算中奖金额
    result.win_amount = CalculateWinAmount(result.grid, bet_amount, config_.active_lines);
    result.profit = result.win_amount - bet_amount;
    
    // 检查免费旋转触发
    if (!in_free_spins) {
        result.trigger_free_spins = CheckFreeSpinsTrigger(result.grid);
        result.free_spins_remaining = result.trigger_free_spins ? config_.free_spins_count : 0;
    } else {
        result.trigger_free_spins = false;
        result.free_spins_remaining = std::max(0, free_spins_remaining - 1);
        // 免费旋转倍数
        result.win_amount *= config_.free_spins_multiplier;
        result.profit = result.win_amount - bet_amount;
    }
    
    return result;
}

void SlotMachine::ResetState() {
    // 老虎机是无状态的，这里主要是为了接口一致性
    // 如果将来需要维护一些状态，可以在这里重置
}

const BetOptions& SlotMachine::GetBetOptions(const std::string& currency) const {
    auto it = config_.bet_table.find(currency);
    if (it != config_.bet_table.end()) {
        return it->second;
    }
    return empty_bet_options_;
}

bool SlotMachine::IsValidBet(float bet_amount, const std::string& currency) const {
    const auto& bet_options = GetBetOptions(currency);
    return std::find(bet_options.begin(), bet_options.end(), bet_amount) != bet_options.end();
}

SpinGrid SlotMachine::GenerateSpinGrid(const std::string& reel_set_name) {
    auto it = reel_sets_.find(reel_set_name);
    if (it == reel_sets_.end()) {
        throw std::runtime_error("Reel set not found: " + reel_set_name);
    }
    
    return it->second->GenerateGrid(rng_);
}

float SlotMachine::CalculateWinAmount(const SpinGrid& grid, float bet_amount, int active_lines) {
    return pay_table_->CalculateTotalWin(grid, bet_amount, active_lines);
}

bool SlotMachine::CheckFreeSpinsTrigger(const SpinGrid& grid) {
    // 检查是否有3个或更多scatter符号在不同列上
    int num_reels = grid.size() / config_.window_size;
    int scatter_columns = 0;
    
    for (int col = 0; col < num_reels; ++col) {
        bool has_scatter = false;
        for (int row = 0; row < config_.window_size; ++row) {
            if (grid[row * num_reels + col] == config_.scatter_symbol) {
                has_scatter = true;
                break;
            }
        }
        if (has_scatter) {
            scatter_columns++;
        }
    }
    
    return scatter_columns >= 3;
}

std::vector<int> SlotMachine::GetPaylineSymbols(const SpinGrid& grid, const PaylineIndices& payline) {
    std::vector<int> symbols;
    symbols.reserve(payline.size());
    
    for (int index : payline) {
        if (index >= 0 && index < static_cast<int>(grid.size())) {
            symbols.push_back(grid[index]);
        }
    }
    
    return symbols;
}

int SlotMachine::CountConsecutiveSymbols(const std::vector<int>& symbols) {
    if (symbols.empty()) return 0;
    
    int first_symbol = symbols[0];
    int count = 1;
    
    for (size_t i = 1; i < symbols.size(); ++i) {
        int current_symbol = symbols[i];
        if (current_symbol == first_symbol || IsWildSymbol(current_symbol)) {
            count++;
        } else if (IsWildSymbol(first_symbol)) {
            // 如果第一个是wild，更新为当前非wild符号
            first_symbol = current_symbol;
            count++;
        } else {
            break;
        }
    }
    
    return count;
}

bool SlotMachine::IsWildSymbol(int symbol) const {
    return std::find(config_.wild_symbols.begin(), config_.wild_symbols.end(), symbol) 
           != config_.wild_symbols.end();
}

} // namespace SlotSimulator